src/coreclr/jit/optimizer.cpp

```diff
@@ -4278,40 +4294,65 @@ void Compiler::optInvertWhileLoop(BasicBlock* block)
         }
     }
 
-    // If the compare has too high cost then we don't want to dup
+    // If the compare has too high cost then we don't want to dup.
 
     bool costIsTooHigh = (estDupCostSz > maxDupCostSz);
 
-    int countOfHelpers = 0;
+    OptInvertCountTreeInfoType optInvertTotalInfo = {};
     if (costIsTooHigh)
     {
         // If we already know that the cost is acceptable, then don't waste time walking the tree
-        // counting shared static helpers.
+        // counting things to boost the maximum allowed cost.
         //
         // If the loop condition has a shared static helper, we really want this loop converted
         // as not converting the loop will disable loop hoisting, meaning the shared helper will
         // be executed on every loop iteration.
-        fgWalkTreePre(&condTree, CountSharedStaticHelper, &countOfHelpers);
+        //
+        // If the condition has array.Length operations, also boost, as they are likely to be CSE'd.
 
-        if (countOfHelpers > 0)
+        for (Statement* stmt : bTest->Statements())
         {
-            maxDupCostSz += 24 * min(countOfHelpers, (int)(loopIterations + 1.5));
+            GenTree* tree = stmt->GetRootNode();
+
+            OptInvertCountTreeInfoType optInvertInfo = {};
+            fgWalkTreePre(&tree, Compiler::optInvertCountTreeInfo, &optInvertInfo);
+            optInvertTotalInfo.sharedStaticHelperCount += optInvertInfo.sharedStaticHelperCount;
+            optInvertTotalInfo.arrayLengthCount += optInvertInfo.arrayLengthCount;
 
-            // Is the cost too high now?
-            costIsTooHigh = (estDupCostSz > maxDupCostSz);
+            if ((optInvertInfo.sharedStaticHelperCount > 0) || (optInvertTotalInfo.arrayLengthCount > 0))
+            {
+                // Calculate a new maximum cost. We might be able to early exit.
+
+                unsigned newMaxDupCostSz =
+                    maxDupCostSz + 24 * min(optInvertTotalInfo.sharedStaticHelperCount, (int)(loopIterations + 1.5)) +
+                    8 * optInvertTotalInfo.arrayLengthCount;
```

Comment:
This formula seems a bit odd to me... 
* not clear why `loopIterations` factors in like it does, and adding `1.5` seems awfully arbitrary
* I would expect this loop to reduce the cost estimate rather than increase the maximum we're willing to pay

Not suggesting you change it yet, mind you, but here's a case where it would be nice to keep track of cost and benefit; having an expensive CSE should both reduce cost (less code ends up getting duplicated) and increase benefit (an expensive computation happens just once, not repeatedly).

Reply:
I didn't follow, nor I see any documentation around the numbers `24`, `8` and `1.5`. Probably worth mentioning it somewhere and use `const` variable with appropriate name to describe it?

Reply:
The calculations here, and numbers chosen, are seem somewhat arbitrary. I don't have a good sense as to *why* they are good choices. The only changes I made to this set of numbers and calculations is:
1. Bumped initial `maxDupCostSz` from 32 to 34. This was specifically to avoid regressions where before we measured the cost of the JTRUE op1, and now we measure the cost of the JTRUE tree itself, which adds a cost of 2.
2. I added `8 * arrayLengthCount` to the max cost, based on appearances of array.Length expressions. This itself was chosen arbitrarily.

So, I don't have a good sense as to why the original numbers of 32, `*4`, `*2`, 12.0, 96.0, 24, or 1.5 were chosen, or why `loopIterations` is used as it is.

Reply:
> I would expect this loop to reduce the cost estimate rather than increase the maximum we're willing to pay

It does seem like that would make sense, and make the code easier to reason about. Since I was sure how to think about the existing sharedStaticHelperCount calculation, I wasn't sure how to translate to that form.